### PR TITLE
Add basic PHPStan linter

### LINKED ISSUE #618: Please use PHPStan
### Describe the bug

There are nasty things in the code, e.g. `return parent::__construct();`

### Steps to Reproduce

The simplest way.
```
composer require --dev szepeviktor/phpstan-wordpress
vendor/bin/phpstan analyze -c vendor/szepeviktor/phpstan-wordpress/extension.neon includes/ providers/ class-*.php two-factor.php --level=0
```

### PR DESCRIPTION
**Pull request by kasparsd**
## What?

Introduces [phpstan](https://phpstan.org) with WordPress core stubs through [szepeviktor/phpstan-wordpress](https://github.com/szepeviktor/phpstan-wordpress).

## Why?

Fixes #618.

Static analysis helps us uncover potential bugs and edge cases. Introducing the tooling didn't require a lot adjustments thanks to work already done in #619.

## How?

Add the tooling as Composer development dependency and introduce additional lint scripts in `composer.json` and `package.json`.

## Testing Instructions

1. Install composer dependencies via `composer install` after checking out this branch.
2. Run `composer lint-phpstan` and confirm that `[OK] No errors` is returned.

## Changelog Entry

Development: Introduce PHPstan tooling for development.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -5,3 +5,4 @@
 /dist/
 /tests/logs/
 .phpunit.result.cache
+phpstan.neon
```
Comment (kasparsd) [inline .gitignore:8]: Allow local overrides.

File: class-two-factor-core.php [hunk 2]
```diff
@@ -93,7 +93,7 @@ class Two_Factor_Core {
 	 * @since 0.1-dev
 	 */
 	public static function add_hooks( $compat ) {
-		add_action( 'init', array( __CLASS__, 'get_providers' ) );
+		add_action( 'init', array( __CLASS__, 'get_providers' ) ); // @phpstan-ignore return.void
```
Comment (kasparsd) [inline class-two-factor-core.php:96]: @szepeviktor The default phpstan table report doesn't include the actual error identifiers making it hard to ignore them. Is there an output format that includes them?

File: composer.json [hunk 3]
```diff
@@ -31,12 +31,14 @@
...
     "phpunit/phpunit": "^8.5|^9.6",
     "spatie/phpunit-watcher": "^1.23",
+    "szepeviktor/phpstan-wordpress": "^1.3",
     "wp-coding-standards/wpcs": "^3.1",
     "yoast/phpunit-polyfills": "^2.0"
   },
   "scripts": {
     "lint": "phpcs",
     "lint-compat": "phpcs -p --standard=PHPCompatibilityWP --runtime-set testVersion 7.2- --extensions=php --ignore='tests/,dist/,includes/Yubico/,vendor/,node_modules/' .",
+    "lint-phpstan": "phpstan analyse --memory-limit=1G",
```
Comment (kasparsd) [inline composer.json:41]: Running without bumping the memory limit can sometimes return passing checks when they actually didn't run.

File: class-two-factor-core.php (hunk 2 above)
Comment (szepeviktor) [reply class-two-factor-core.php:96]: Yes. Here is an example
https://github.com/TicketSwap/phpstan-error-formatter

Comment (szepeviktor) [reply class-two-factor-core.php:96]: but there is https://github.com/phpstan/phpstan-src/pull/3112
https://phpstan.org/user-guide/output-format

Comment (szepeviktor) [reply class-two-factor-core.php:96]: `raw` should also display it

Comment (kasparsd) [reply class-two-factor-core.php:96]: Unfortunately, it doesn't:

no-error-ident

Comment (szepeviktor) [reply class-two-factor-core.php:96]: try adding a `-v`
https://github.com/phpstan/phpstan-src/blob/62c81bac66f2558c2768914d3a79abad7c4c8b3d/src/Command/ErrorFormatter/RawErrorFormatter.php#L22

Comment (kasparsd) [reply class-two-factor-core.php:96]: That worked also for the default output. I'm going to add that in as the default in the `composer.json` script.

verbose

**Comment by szepeviktor** [comment]
Is phpstan output on the current level empty?

**Review by dd32 — APPROVED** [review]
LGTM

**Pull request merged**